orderbook/src/fee.rs

```diff
@@ -0,0 +1,180 @@
+use std::collections::HashMap;
+
+use anyhow::Result;
+use chrono::{DateTime, FixedOffset, Utc};
+use primitive_types::{H160, U256};
+use std::sync::Mutex;
+
+use crate::price_estimate::PriceEstimating;
+use gas_estimation::GasPriceEstimating;
+
+type Measurement = (U256, DateTime<Utc>);
+
+pub struct MinFeeCalculator {
+    price_estimator: Box<dyn PriceEstimating>,
+    gas_estimator: Box<dyn GasPriceEstimating>,
+    native_token: H160,
+    // TODO persist past measurements to shared storage
+    measurements: Mutex<HashMap<H160, Vec<Measurement>>>,
+    now: Box<dyn Fn() -> DateTime<Utc>>,
+}
+
+const GAS_PER_ORDER: f64 = 100_000.0;
+const STANDARD_VALIDITY_FOR_FEE_IN_SEC: i32 = 60;
+
+impl MinFeeCalculator {
+    pub fn new(
+        price_estimator: Box<dyn PriceEstimating>,
+        gas_estimator: Box<dyn GasPriceEstimating>,
+        native_token: H160,
+    ) -> Self {
+        Self {
+            price_estimator,
+            gas_estimator,
+            native_token,
+            measurements: Default::default(),
+            now: Box::new(Utc::now),
+        }
+    }
+}
+
+impl MinFeeCalculator {
+    // Returns the minimum amount of fee required to accept an order selling the specified token
+    // and an expiry date for the estimate.
+    pub async fn min_fee(&self, token: H160) -> Result<Measurement> {
+        let gas_price = self.gas_estimator.estimate().await?;
+        let token_price = self
+            .price_estimator
+            .estimate_price(token, self.native_token)
+            .await?;
+
+        let min_fee = U256::from_f64_lossy(gas_price * token_price * GAS_PER_ORDER);
+        let valid_until = (self.now)() + FixedOffset::east(STANDARD_VALIDITY_FOR_FEE_IN_SEC);
+        let result = (min_fee, valid_until);
+
+        self.measurements
+            .lock()
+            .expect("Thread holding mutex panicked")
+            .entry(token)
+            .or_default()
+            .push(result);
+        Ok(result)
+    }
+
+    // Returns true if the fee satisfies a previous not yet expired estimate, or the fee is high enough given the current estimate.
+    pub async fn is_valid_fee(&self, token: H160, fee: U256) -> bool {
+        if let Some(measurements) = self
+            .measurements
+            .lock()
+            .expect("Thread holding mutex panicked")
+            .get(&token)
+        {
+            // Previous measurements are ordered ascending by expiry data
+            for (suggested_fee, expiry_date) in measurements.iter().rev() {
+                if expiry_date < &(self.now)() {
+                    break;
+                }
```

Comment:
Maybe this could also remove estimates that have expired? Otherwise we accumulate them forever. But if you're changing this component soon anyway then not important.